File: 算法/代码随想录算法PDF/00调试源代码/216.组合总和-iii.cpp
/*
 * @lc app=leetcode.cn id=216 lang=cpp
 *
 * [216] 组合总和 III
 */

// @lc code=start
class Solution {
public:
    int globalN, globalK;
    vector<int> path;
    vector<vector<int>> res;
    void backtracking(int need, int startIndex) {
        if (need < 0) return ;
        if (path.size() == globalK) {
            if (need == 0) res.push_back(path);
            return ;
        }
        for (int i = startIndex; i <= 9 - (globalK - path.size()) + 1; i++) {
            path.push_back(i);
            backtracking(need - i, i + 1);
            path.pop_back();
        }
    }
    vector<vector<int>> combinationSum3(int k, int n) {
        globalN = n;
        globalK = k;
        backtracking(n, 1);
        return res;
    }
};
// @lc code=end

File: 算法/代码随想录算法PDF/00调试源代码/216.组合总和-iii.py
#
# @lc app=leetcode.cn id=216 lang=python3
#
# [216] 组合总和 III
#

# @lc code=start
class Solution:
    def combinationSum3(self, k: int, n: int) -> List[List[int]]:
        res = []
        path = []
        def backtracking(need, startIndex):
            nonlocal path, res, n, k
            if need < 0: return  # 剪枝
            if len(path) == k:
                if need == 0:
                    res.append(path[:])
                return
            for i in range(startIndex, 10 - (k - len(path)) + 1):
                path.append(i)
                backtracking(need - i, i + 1)
                path.pop()
        
        backtracking(n, 1)
        return res
# @lc code=end

